File: my-submissions/0034-find-first-and-last-position-of-element-in-sorted-array/solution.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> res = {-1,-1};
        int lb = lowerbound(nums,target);
        if(lb == nums.size()|| nums[lb] != target )
        return res;
        int ub = upperbound(nums,target) -1;
        res[0] = lb;
        res[1] = ub;
        return res;
    }

    int lowerbound(vector<int> &nums,int tar){
        
     return lower_bound(nums.begin(),nums.end(),tar) - nums.begin();
        
    }
    int upperbound(vector<int> &nums,int tar){
        
       return   upper_bound(nums.begin(),nums.end(),tar) - nums.begin();
        
    }
};

File: my-submissions/0034-find-first-and-last-position-of-element-in-sorted-array/solution.java
class Solution {
    static int lower_bound(int[] nums,int tar){
        int low = 0;
        int high = nums.length - 1;

        int ans = nums.length;

        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] >= tar){
                ans = mid;
                high = mid - 1;
            }
            else{
                low = mid +1;
            }
        }
        return ans;
    }

    static int higher_bound(int[] arr,int tar){
        int low = 0;
        int high = arr.length - 1;
        int ans = arr.length;

        while(low <= high){
            int mid = low + (high - low)/2;

            if(arr[mid] <= tar){
                ans = mid;
                low = mid + 1;
            }
            else    high = mid -1;
        }
        return ans;
    }
    
    public int[] searchRange(int[] nums, int target) {
        int first = lower_bound(nums,target);
        if(first == nums.length || nums[first] != target) return new int[] {-1,-1};
        int last = higher_bound(nums,target);

        return new int[] {first, last};

    }
}
